go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,168 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	ipam "k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
+
+	"reflect"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"k8s.io/klog"
+)
+
+// logicalSwitchInfo contains information corresponding to the node. It holds the
+// subnet allocations (v4 and v6) as well as the IPAM allocator instances for each
+// subnet managed for this node
+// logicalSwitchInfo is used to save switch specific information such as subnets,
+// allocated IPs per switch etc.
+type logicalSwitchInfo struct {
+	hostSubnets []*net.IPNet
+	ipams       []ipam.Interface
+}
+
+type logicalSwitchCache struct {
+	cache map[string]logicalSwitchInfo
+	// A RW mutex for logicalSwitchCache which holds logicalSwitch information
+	sync.RWMutex
+}
+
+type ipamFactoryFunc func(*net.IPNet) ipam.Interface
+
+func newIPAMAllocator(cidr *net.IPNet) ipam.Interface {
+	return ipam.NewCIDRRange(cidr)
+}
+
+func (cache *logicalSwitchCache) Add(nodeName string, hostSubnets []*net.IPNet, ipamFunc ipamFactoryFunc) error {
+	cache.Lock()
```

Comment:
Yes, because the Add is serialized by the `Lock()`

Reply:
On the lock issue my question was a bit different.

node-name and subnet is  passed into the Add function and  we're reserving ip's from that subnet. So my question was regarding how long we're holding the write lock

right now we're holding it for entire duration of the Add function. So my question is since we are operating on the data being passed into the function and I assume subnet allocation per node is already done on the higher layer, should we hold the lock for the entire duration of the Add method? Or should we just hold it while updating the cache. I was asking this because, with the write lock  held no one else can do anything with the object including reads. Especially considering this as a performance feature, question was should we hold the lock only while writing into the cache(map) object Or is there a reason we are holding it for the entire duration of add method making everyone else wait?

Thanks for discussing this with me. As per discussion with you this is not an issue as the operation should be infrequent. Just adding the notes here for concluding this topic.